File: src/main/java/jb/validator/SimpleScoreValidator.java
package jb.validator;

import de.bwaldvogel.liblinear.Model;
import jb.config.Opts;
import jb.data.Problem;
import jb.data.ScoreTuple;
import jb.data.ValidatingTestingTuple;
import jb.data.clfobj.ClfObjectOnceSorted;

import java.util.List;

import static jb.util.ModelUtils.predictsPropperly;


public class SimpleScoreValidator implements Validator {

    @Override
    public ScoreTuple validate(List<? extends Model> clfs, ValidatingTestingTuple validatingTestingTuple, Opts opts) {

        double[][] scores = new double[clfs.size()][opts.getNumberOfSpaceParts()];
        double[][] weights = new double[opts.getNumberOfSpaceParts()][clfs.size()];


        for (int i = 0; i < clfs.size(); i++) {
            for (int j = 0; j < opts.getNumberOfSpaceParts(); j++) {
                double propperlyClassified = 0;
                Problem currentProblem = validatingTestingTuple.getValidationProblems().get(j);
                for (int k = 0; k < currentProblem.getX().length; k++) {
                    if (predictsPropperly(clfs.get(i), new ClfObjectOnceSorted(currentProblem.getX()[k], currentProblem.getY()[k]))) {
                        propperlyClassified++;
                    }
                }
                scores[i][j] = propperlyClassified / validatingTestingTuple.getValidationProblems().get(j).getY().length;
                weights[j][i] = propperlyClassified / validatingTestingTuple.getValidationProblems().get(j).getY().length;
            }
        }

        return new ScoreTuple(scores, weights);
    }

}

File: src/main/java/jb/validator/SimpleScoreValidator.kt
package jb.validator

import de.bwaldvogel.liblinear.Model
import jb.config.Opts
import jb.data.Problem
import jb.data.ScoreTuple
import jb.data.ValidatingTestingTuple
import jb.data.clfobj.ClfObjectOnceSorted

import jb.util.ModelUtils.predictsPropperly


class SimpleScoreValidator : Validator {

    override fun validate(clfs: List<Model>, validatingTestingTuple: ValidatingTestingTuple, opts: Opts): ScoreTuple {

        val scores = Array(clfs.size) { DoubleArray(opts.numberOfSpaceParts) }
        val weights = Array(opts.numberOfSpaceParts) { DoubleArray(clfs.size) }


        for (i in clfs.indices) {
            for (j in 0 until opts.numberOfSpaceParts) {
                var propperlyClassified = 0.0
                val (x, y) = validatingTestingTuple.validationProblems[j]
                for (k in 0 until x.size) {
                    if (predictsPropperly(clfs[i], ClfObjectOnceSorted(x[k], y[k]))) {
                        propperlyClassified++
                    }
                }
                scores[i][j] = propperlyClassified / validatingTestingTuple.validationProblems[j].y.size
                weights[j][i] = propperlyClassified / validatingTestingTuple.validationProblems[j].y.size
            }
        }

        return ScoreTuple(scores, weights)
    }

}
